lld/MachO/ObjC.cpp

```diff
@@ -320,3 +332,915 @@ void objc::checkCategories() {
+  // would provide the same info)
+  if (!infoCategoryWriter.catPtrListInfo.valid) {
+    for (uint32_t off = catLayout.instanceMethodsOffset;
+         off <= catLayout.classPropsOffset; off += target->wordSize) {
+      Defined *ptrList;
+      if (tryGetDefinedAtIsecOffset(catInfo.catBodyIsec, off, ptrList)) {
+        collectSectionWriteInfoFromIsec<ConcatOutputSection>(
+            ptrList->isec, infoCategoryWriter.catPtrListInfo);
+        // we've successfully collected data, so we can break
+        break;
+      }
+    }
+  }
+}
+
+// Parse a protocol list that might be linked to at a ConcatInputSection given
+// offset. The format of the protocol list is different than other lists (prop
+// lists, method lists) so we need to parse it differently
+void ObjcCategoryMerger::parseProtocolListInfo(ConcatInputSection *isec,
+                                               uint32_t secOffset,
+                                               PointerListInfo &ptrList) {
+  if (!isec || (secOffset + target->wordSize > isec->data.size()))
+    assert("Tried to read pointer list beyond protocol section end");
+
+  const Reloc *reloc = isec->getRelocAt(secOffset);
+  if (!reloc)
+    return; // List is null, nothing to do
+
+  auto *ptrListSym = dyn_cast_or_null<Defined>(reloc->referent.get<Symbol *>());
+  assert(ptrListSym && "Protocol list reloc does not have a valid Defined");
+
+  // Theoretically protocol count can be either 32b or 64b, but reading the
+  // first 32b is good enough
+  uint32_t protocolCount = *reinterpret_cast<const uint32_t *>(
+      ptrListSym->isec->data.data() + listHeaderLayout.structSizeOffset);
+
+  ptrList.structCount += protocolCount;
+  ptrList.structSize = target->wordSize;
+
+  uint32_t expectedListSize =
+      (protocolCount * target->wordSize) +
+      /*header(count)*/ protocolListHeaderLayout.totalSize +
+      /*extra null value*/ target->wordSize;
+  assert(expectedListSize == ptrListSym->isec->data.size() &&
+         "Protocol list does not match expected size");
+
+  uint32_t off = protocolListHeaderLayout.totalSize;
+  for (uint32_t inx = 0; inx < protocolCount; inx++) {
+    const Reloc *reloc = ptrListSym->isec->getRelocAt(off);
+    assert(reloc && "No reloc found at protocol list offset");
+
+    auto *listSym = dyn_cast_or_null<Defined>(reloc->referent.get<Symbol *>());
+    assert(listSym && "Protocol list reloc does not have a valid Defined");
+
+    ptrList.allPtrs.push_back(listSym);
+    off += target->wordSize;
+  }
+}
+
+// Parse a pointer list that might be linked to at a ConcatInputSection given
+// offset. This can be used for instance methods, class methods, instance props
+// and class props since they have the same format.
+void ObjcCategoryMerger::parsePointerListInfo(ConcatInputSection *isec,
+                                              uint32_t secOffset,
+                                              uint32_t symbolsPerStruct,
```

Comment:
I will add an assert that verifies integrity. Is it OK to have it just for assert ?

Reply:
It's odd to add a parameter just for an assert. why not checking it at the call-site?

Reply:
The function reads the header info and ensures that the `symbolsPerStruct` matches what was read.

Reply:
Can we use an enum to handle 4 each case -- instance methods, class methods, instance props
 and class props, and use it to specialize code inside this function, instead of individually passing offset, size, and output value, etc, which are error-prone?
 Basically, you could attach `extInfo` to `ObjcCategoryMerger`, which you can avoid passing output parameters all over the place, and call something like:
 ```
 parsePointerListInfo(catInfo.catBodyIsec, InstanceMethods);
 parsePointerListInfo(catInfo.catBodyIsec, ClassMethods);
 parsePointerListInfo(catInfo.catBodyIsec, InstanceProps);
 parsePointerListInfo(catInfo.catBodyIsec, ClassProps);
 ```

Reply:
Great idea - but it will only work for symbolsPerStruct. I'll move that into `PointerListInfo`. 
For the offset - it would work now, but won't work when extending to also parsing classes. The offsets there will be different so we can't incorporate them into `PointerListInfo`.

Actually, I now realize I can't incorporate `pszSymNamePrefix` into `PointerListInfo` info also -since that will also differ when classes are also handeled. I'll also move `pszSymNamePrefix` out of `PointerListInfo`.

Reply:
Another option is to have 
```
PointerListInfo {
  char *classPrefix;
  char *categoryPrefix;
  
  uint32_t classOffset;
  uint32_t categoryOffset;
}
```
and do :
```
parsePointerListInfo(catInfo.catBodyIsec, InstanceMethods, /*fromClass*/=false);
```
otherwise we have to do:
```
parsePointerListInfo(catInfo.catBodyIsec, InstanceMethods, offset, symPrefix);
```